### PR TITLE
Coverage

### PR DESCRIPTION
**Pull request by outscale-toa**

### CODE REVIEW

File: scripts/tests_coverage.sh [hunk 1]
```diff
@@ -0,0 +1,55 @@
+#!/bin/sh
+
+# Butterfly build root
+BUTTERFLY_BUILD_ROOT=$1
+
+# Test Butterfly build root
+if [ ! -d $BUTTERFLY_BUILD_ROOT/api ]; then
```
Comment (outscale-jju) [inline scripts/tests_coverage.sh]: warning, someone putting butterfly's source root will pass this test as api folder also exist.

File: scripts/tests_coverage.sh [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    echo -e "
+Coverage files (*.gcda) not found
+in $client_obj_dir 
+or 
+in $server_obj_dir
+Make sure you did that in your build directory:
+- cmake .. -DDEBUG=ON
+- make 
+run some tests before running coverage test.
+"
```
Comment (outscale-jju) [inline scripts/tests_coverage.sh]: that's weird, please use one-line echo and split in several echo (one per line)

File: scripts/tests_coverage.sh [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+server_obj_dir=$BUTTERFLY_BUILD_ROOT/api/server/CMakeFiles/api_server.dir/
+
+# Test coverage data files
+ret=$(find $BUTTERFLY_BUILD_ROOT/api/ -name *.gcda | wc -l)
+if [ $ret -eq 0 ]; then
+    echo -e "
+Coverage files (*.gcda) not found
+in $client_obj_dir 
+or 
+in $server_obj_dir
```
Comment (outscale-jju) [inline scripts/tests_coverage.sh]: in X or in X ... could remove one ?

File: scripts/tests_coverage.sh [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+- make 
+run some tests before running coverage test.
+"
+    exit 1
+fi
+
+function check_error {
+output=$1
+if [ $output -ne 0 ]; then
+    echo COVERAGE TEST FAILED!!
```
Comment (outscale-jju) [inline scripts/tests_coverage.sh]: no need to scream :)

File: scripts/tests_coverage.sh [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    exit 1
+fi
+}
+
+function clean_coverage_data {
+    find $BUTTERFLY_BUILD_ROOT/api/ \( -name *.gcda \) -exec rm {} \;
```
Comment (outscale-jju) [inline scripts/tests_coverage.sh]: why the () ?

File: scripts/tests_api.sh [hunk 6]
```diff
@@ -43,7 +44,7 @@ for request in $BUTTERFLY_ROOT/api/tests/*/*_in; do
 done
 
 sudo chown $USER out.txt
-
+sudo chown $USER:$USER -R $output_dir
```
Comment (outscale-jju) [inline scripts/tests_api.sh]:
```
scripts: edit script for coverage test
-->
scripts: add coverage test
```

File: CMakeLists.txt [hunk 7]
```diff
@@ -80,6 +80,10 @@ add_custom_target(tests-scenario
                   COMMAND ${PROJECT_SOURCE_DIR}/scripts/tests_scenario.sh ${PROJECT_BINARY_DIR} ${VERBOSE}
                   WORKING_DIRECTORY ${PROJECT_BINARY_DIR})
 
+add_custom_target(tests-coverage
+                  COMMAND ${PROJECT_SOURCE_DIR}/scripts/tests_coverage.sh ${PROJECT_BINARY_DIR}
+                  WORKING_DIRECTORY ${PROJECT_BINARY_DIR})
+
```
Comment (outscale-jju) [inline CMakeLists.txt]:
```
api: update cmakefiles for coverage test
-->
build: add coverage tests in cmake files
```

File: scripts/tests_coverage.sh [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    check_error $?
+    rm client.info server.info
+    check_error $?
+}
+
+lcov -c -d $client_obj_dir -o client.info
```
Comment (outscale-jju) [inline scripts/tests_coverage.sh]: This binary should be checked before
something like `lcov -h || echo "Error: lcov command not found" && exit 1`

File: api/server/CMakeLists.txt [hunk 9]
```diff
@@ -40,6 +40,15 @@ target_link_libraries(api_server
                       ${PG_LIBRARIES}
                       )
 
-set(CMAKE_CXX_FLAGS "-std=c++11 ${CMAKE_CXX_FLAGS} -Wall -g")
-set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -pthread -static-libgcc -static-libstdc++")
+option(DEBUG "Coverage" OFF)
+if(DEBUG)
+   set(CMAKE_CXX_OUTPUT_EXTENSION_REPLACE 1)
+   set(GCC_COVERAGE_COMPILE_FLAGS  "-g -O0 -fprofile-arcs -ftest-coverage ")
+   set(GCC_COVERAGE_LINK_FLAGS  "-lgcov")
```
Comment (outscale-jju) [inline api/server/CMakeLists.txt]: You are repeating yourself

File: CMakeLists.txt [hunk 10]
```diff
@@ -45,6 +45,11 @@ set(PG_INCLUDE_DIR ${3RDPARTY_DIR}/packetgraph/include)
 
 set(CMAKE_C_FLAGS "-g -O3 -march=core-avx-i -mtune=core-avx-i -fmessage-length=0  -Werror -Wall -Wextra -Wwrite-strings -Winit-self -Wcast-align -Wpointer-arith -Wstrict-aliasing -Wformat=2 -Wmissing-declarations -Wmissing-include-dirs -Wno-unused-parameter -Wuninitialized -Wold-style-definition -Wstrict-prototypes -Wmissing-prototypes -L${DPDK_INSTALL_DIR}/build/lib")
 
+option(DEBUG "Coverage" OFF)
```
Comment (outscale-mgo) [inline CMakeLists.txt]: if DEBUG is set when you need Coverage, DEBUG should be rename to... COVERAGE

File: api/server/CMakeLists.txt [hunk 11]
```diff
@@ -40,6 +40,11 @@ target_link_libraries(api_server
                       ${PG_LIBRARIES}
                       )
 
-set(CMAKE_CXX_FLAGS "-std=c++11 ${CMAKE_CXX_FLAGS} -Wall -g")
-set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -pthread -static-libgcc -static-libstdc++")
+if(DEBUG)
```
Comment (outscale-mgo) [inline api/server/CMakeLists.txt]: s/DEBUG/COVERAGE/

**Comment by outscale-mgo** [comment]
As you use tests_coverage.sh in CMakeList.txt, the patch that add tests_coverage.sh must be before the one touching about cmake, or you will have a patch that break everything.

File: scripts/tests_coverage.sh [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+# Butterfly build root
+BUTTERFLY_BUILD_ROOT=$1
+
+client_obj_dir=$BUTTERFLY_BUILD_ROOT/api/client/CMakeFiles/api_client.dir/
+server_obj_dir=$BUTTERFLY_BUILD_ROOT/api/server/CMakeFiles/api_server.dir/
+
+# Test Butterfly build root
+if [ ! -d $client_obj_dir -o ! -d $server_obj_dir ]; then
+    echo "Please set butterfly's build root as parameter"
```
Comment (outscale-mgo) [inline scripts/tests_coverage.sh]: echo "either '$client_obj_dir' or '$server_obj_dir' is not a directory have you set BUTTERFLY_BUILD_ROOT ?"
would be more explicit

File: scripts/tests_coverage.sh [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+if [ $ret -eq 0 ]; then
+    echo " Coverage files (*.gcda) not found in $client_obj_dir"
+    echo " Make sure you did that in your build directory:"
+    echo " - cmake .. -DDEBUG=ON"
+    echo " - make"
+    echo " run some tests before running coverage test."
+    exit 1
+fi
+
+rm -fr $BUTTERFLY_BUILD_ROOT/client_coverage
```
Comment (outscale-mgo) [inline scripts/tests_coverage.sh]: I generally use -rfv when I'm using an rm, like this if something goes wrong in this test, the user might notice it before it's too late

File: scripts/tests_coverage.sh [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+rm -fr $BUTTERFLY_BUILD_ROOT/server_coverage
+
+function check_error {
+output=$1
+if [ $output -ne 0 ]; then
```
Comment (outscale-mgo) [inline scripts/tests_coverage.sh]: s/$output/$1/
and you can remove 1 line :)
don't forget less code is better

File: tests/functions.sh [hunk 15]
```diff
@@ -1042,6 +1042,7 @@ function clean_all {
     sudo killall -9 butterflyd butterfly qemu-system-x86_64 socat &> /dev/null || true
     sudo rm -rf /tmp/*vhost* /dev/hugepages/* /mnt/huge/*  &> /dev/null
     sleep 0.5
+    sudo chown $USER:$USER -R $BUTTERFLY_BUILD_ROOT/api/
```
Comment (outscale-mgo) [inline tests/functions.sh]: can't you move this to coverage test ?

**Comment by outscale-jju** [comment]
made some fixes and merged

**Pull request closed without merging**